drivers/flash/flash_stm32.c

```diff
@@ -210,11 +210,9 @@ static int flash_stm32_write_protection(struct device *dev, bool enable)
 	} else {
 		if (regs->cr & FLASH_CR_LOCK) {
 			regs->keyr = FLASH_KEY1;
-#ifdef CONFIG_SOC_SERIES_STM32F3X
-			/* On STM32F3 series some time is requested */
-			/* before writing again on key register */
-			flash_stm32_wait_flash_idle(dev);
-#endif /* CONFIG_SOC_SERIES_STM32F3X */
+			/* Order of operations is critical here */
+			/* Add a barrier to avoid compiler optimisations */
+			__ISB();
 			regs->keyr = FLASH_KEY2;
```

Comment:
I am not sure that you need an instruction barrier, the flash registers are in a strongly ordered region, so the CPU should already ensure that the two writes actually reach the registers.

OTOH you might want a compiler barrier, but that should not be needed if the registers are declared `volatile`. And I just realized that all the flash registers are declared as volatile, except on the STM32F1x and STM32F3x. Could it be the real issue?

Reply:
>  Could it be the real issue?

This sounds more realistic indeed. I'll have a try.

Reply:
That was it, thanks for the hint!